### ISSUE TITLE
Error in the shader of POM

### ISSUE THREAD

**Issue by galek** [JoeyDeVries/LearnOpenGL#19]
Hi

I found problem in shader for POM and CSM

In this line, driver generated TDM error:
float numLayers = mix(maxLayers, minLayers, abs(dot(vec3(0.0, 0.0, 1.0), viewDir)));

**Comment by JoeyDeVries** [comment]
What's the specific error message your driver generated?

**Comment by galek** [comment]
http://nvidia.custhelp.com/app/answers/detail/a_id/3633
http://images.gameru.net/image/direct/1dc2c1fef6.png

**Comment by JoeyDeVries** [comment]
Sorry for the delay. I assume you're talking about a TDR issue as what you've linked and not a TMD issue you've originally posted? 
My guess is that these errors are caused by enormous shaders or infinite loops in a shader program. Seeing as the line you posted is a simple lerp calculation I'd say that probably isn't the cause of the error. Could you perhaps try and comment certain statements to see if you still get the error and if so, pinpoint its location?

**Comment by galek** [comment]
I found, in this fragment

```javascript
while (currentLayerDepth < currentDepthMapValue)
	{
		// shift texture coordinates along direction of P
		currentTexCoords -= deltaTexCoords;
		// get depthmap value at current texture coordinates
		currentDepthMapValue = texture(u_texture_1, currentTexCoords).r;
		// get depth of next layer
		currentLayerDepth += layerDepth;
	}
```

**Comment by galek** [comment]
yeah, i found what step parallax is correct for me, POM i will test later

I found what correct result for simple parallax:
```javascript
float height = (1.0 - texture(normalMap, texCoords).r);
	return texCoords -= E_TBN.xy / E_TBN.z * (height * height_scale);
```

**Comment by galek** [comment]
I found a more precise location:
it's
```javascript
 abs(dot(vec3(0.0, 0.0, 1.0), viewDir))
```
if viewDir.z = 0, then all be ok
